### ISSUE TITLE
`incoming` branch fails with Suite

### ISSUE THREAD

**Issue by tsusanka** [trezor/trezor-user-env#72]
After merging #58 I have restarted Suite e2e job which [runs against](https://gitlab.com/satoshilabs/trezor/trezor-suite/-/commit/3fce6f8d48197d12d1aa279eaaef703a33bfb5ed) the incoming docker image and [it fails](https://gitlab.com/satoshilabs/trezor/trezor-suite/-/jobs/1414462421):

```
CypressError: `cy.task('stopBridge')` failed with the following error:
> websocket_error_message: RuntimeError - Bridge is not running, cannot be stopped
Because this error occurred during a `before each` hook we are skipping all of the remaining tests.
```

So this is mostly likely caused by the fact that Bridge was not started properly.

What is a bit unpleasant is that the CI job does not print trezor-user-env log. @vdovhanych can you get that for us or do we need to run it manually? @grdddj could you please coordinate with @vdovhanych?

Also I have created an issue https://github.com/satoshilabs/devops/issues/37 to have the log as an artifact (@grdddj you need to accept an invite to see it).

**Comment by grdddj** [comment]
I tried to identify the place that causes the issue, and it seems we are trying to stop the bridge and emulator before every test as a cleanup (even if they do not run)

![image](https://user-images.githubusercontent.com/42543243/125190959-42b6e980-e240-11eb-9897-85ad2eccf2fa.png)

We have actually implemented the `cleanup` argument in our code (`bridge.py`), which when True, will not raise that RuntimeError

![image](https://user-images.githubusercontent.com/42543243/125191010-86115800-e240-11eb-8c4f-33ea6ae5739b.png)

Possible solutions (starting with the easiest one):
- just removing the RuntimeError in our code and add a log statement there (like `WARNING: bridge is stopped but is not running`)
- implementing new function in both suite and trezor-user-env, something like `make_sure_bridge(emulator)_is_stopped`, that can be run before these tests

**Comment by tsusanka** [comment]
I think stopping Bridge even if it is not running is a valid call. So I would do your first suggestion.

**Comment by tsusanka** [comment]
Let's try #75 because I think it is the right thing to do anyway but https://github.com/trezor/trezor-user-env/issues/53#issuecomment-877821922 made me think this might not be the reason the CI fails. The Bridge should be running when running `stop` I guess so we should investigate if the Bridge was indeed spawned.

**Comment by grdddj** [comment]
> I guess so we should investigate if the Bridge was indeed spawned

Definitely from the logs in CI (when we have them) we should look into that.

> The Bridge should be running when running `stop`

It doesn't need to be true always, as the `stopBridge` command is being run `before` every test suite, and at least before the first one, there is most probably no bridge running (not sure about all others, when the bridge can be already spawned form the previous test)

**Comment by grdddj** [comment]
Found some clue to what is happening (thanks to `websocket_error_message` string in the test output)

When the websocket response contains the `error` key, it will be regarded as error in the Suite CI, and will cause the jobs to fail.

https://github.com/trezor/trezor-suite/blob/0868666b806a04ab22f7bb121e8a537d62717d94/packages/blockchain-link/src/workers/blockbook/websocket.ts#L143-L147

Us removing the RuntimeError will fix this, but there could be some better ways - I will be thinking about it

**Comment by vdovhanych** [comment]
Yes, if something in the CI run will present 1 on error it will fail. But this fail was from the cypress not from the CI (GitLab) itself, but my assumption is that it will behave the same.

**Comment by grdddj** [comment]
You are right, I might have said it badly - the CI tests fail because of the Cypress errors.

Actually that code link above is not the one that processes the websocket responses, it is this one: (seems like github does not show the code from other repository directly in the comment)

https://github.com/trezor/trezor-suite/blob/0868666b806a04ab22f7bb121e8a537d62717d94/packages/integration-tests/projects/suite-web/plugins/websocket-client.js#L118-L141

So the main thing in the response is the `success` boolean, and if False, it will cause the Cypress tests to fail, as it will throw JavaScript error.

Possible solution:
- in cases that `are errors, but not really`, send success: True and just include a warning in the response

**Comment by grdddj** [comment]
in PR #76 there is a quick POC how it could look like - satisfying both the client with all the info but not messing with Suite Cypress tests

**Comment by tsusanka** [comment]
#75 merged. Job [restarted](https://gitlab.com/satoshilabs/trezor/trezor-suite/-/jobs/1416760385), let's see!

**Comment by grdddj** [comment]
> #75 merged. Job [restarted](https://gitlab.com/satoshilabs/trezor/trezor-suite/-/jobs/1416760385), let's see!

So we got the same issue with the `startBridge` function, so I guess let's remove that RuntimeError from there as well (will open a new PR)

```
CypressError: `cy.task('startBridge')` failed with the following error:
> websocket_error_message: RuntimeError - Bridge is already running, not spawning a new one
```

However, it can also be a sign that something with tests is not completely alright - as they try to start bridge when it already runs (but theoretically those tests run in parallel - **would be good to connect with people who write the Cypress tests**)

**Comment by tsusanka** [comment]
Cc @mroz22, he was responsible for many Cypress infrastructure. Maybe some tests are missing stopBridge? Not sure

**Comment by grdddj** [comment]
After merging #77, there are no more RuntimeErrors to be thrown - so we can explore other errors we encounter :)

**Comment by tsusanka** [comment]
Restarted: https://gitlab.com/satoshilabs/trezor/trezor-suite/-/jobs/1417536668

**Comment by grdddj** [comment]
> Restarted: https://gitlab.com/satoshilabs/trezor/trezor-suite/-/jobs/1417536668

Now we have new error:
```
CypressError: `cy.task('setupEmu')` failed with the following error:
> websocket_error_message: OSError - [Errno 98] Address already in use
```

However, this function is also calling the bridge-start/stop ones inside (see below)

https://github.com/trezor/trezor-suite/blob/0868666b806a04ab22f7bb121e8a537d62717d94/packages/integration-tests/projects/suite-web/plugins/index.js#L101-L124

That `OSError - [Errno 98] Address already in use` seems to come from `bridge-start`, which is trying to run the bridge even when it is already running

Conclusion:
- we really need to look at the `master` branch of `trezor-user-env` and see how these start() and stop() functions differ from our current version

**Comment by grdddj** [comment]
>  we really need to look at the `master` branch of `trezor-user-env` and see how these start() and stop() functions differ from our current version

I was not able to find any logical difference in them

@vdovhanych right now having the `debugging.log` file would be really useful. Can you please incorporate it as an artifact into that Suite CI job? (if it is not there already, but I see only suite stuff - videos etc.)

**Comment by grdddj** [comment]
(Possible time saver)
- can I somehow run and troubleshoot these problems locally without having to coordinate and always merge into incoming and run these tests in CI?
- it should be possible to run this Cypress tests locally using this [suite guide](https://github.com/trezor/trezor-suite/blob/develop/docs/tests/e2e-web.md) - do you have any experience trying that (would you recommend it)?

**Comment by vdovhanych** [comment]
I will try to add it to that job and probably do it today evening or tmrw morning. Also, you definitely can run the tests locally. You just need docker, and that's pretty much it. It shouldn't be a problem.

**Comment by tsusanka** [comment]
> it should be possible to run this Cypress tests locally using this suite guide - do you have any experience trying that (would you recommend it)?

Yes this readme should be up to date and working, I was using that just recently.

**Comment by grdddj** [comment]
Nice, thank you for the help, will try that and let you know
